src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteCache.java

```diff
@@ -255,50 +262,118 @@ public void ensureInputsPresent(
+    @Nullable Supplier<HashCode> hashSupplier = null;
+    if (options.remoteVerifyDownloads) {
+      HashingOutputStream hashOut = digestUtil.newHashingOutputStream(out);
+      hashSupplier = hashOut::hash;
+      out = hashOut;
+    }
+
     SettableFuture<Void> outerF = SettableFuture.create();
+    requestRead(resourceName, /* offset=*/ 0, retrier.newBackoff(), digest, out, hashSupplier, outerF);
+    return outerF;
+  }
+
+  private void requestRead(
+      String resourceName,
+      long offset,
+      Backoff backoff,
+      Digest digest,
+      OutputStream out,
+      @Nullable Supplier<HashCode> hashSupplier,
+      SettableFuture<Void> future) {
     bsAsyncStub()
         .read(
-            ReadRequest.newBuilder().setResourceName(resourceName).build(),
+            ReadRequest.newBuilder()
+                .setResourceName(resourceName)
+                .setReadOffset(offset)
+                .build(),
             new StreamObserver<ReadResponse>() {
+              Backoff stalledBackoff = backoff;
+              long committed = 0;
+
               @Override
               public void onNext(ReadResponse readResponse) {
+                ByteString data = readResponse.getData();
                 try {
-                  readResponse.getData().writeTo(hashOut != null ? hashOut : out);
+                  data.writeTo(out);
+                  committed += data.size();
                 } catch (IOException e) {
-                  outerF.setException(e);
+                  future.setException(e);
                   // Cancel the call.
                   throw new RuntimeException(e);
                 }
+                // reset the stall backoff because we've made progress or been kept alive
+                stalledBackoff = retrier.newBackoff();
               }
 
               @Override
               public void onError(Throwable t) {
-                if (t instanceof StatusRuntimeException
-                    && ((StatusRuntimeException) t).getStatus().getCode()
-                        == Status.NOT_FOUND.getCode()) {
-                  outerF.setException(new CacheNotFoundException(digest, digestUtil));
-                } else {
-                  outerF.setException(t);
+                Status status = Status.fromThrowable(t);
+                boolean useBackoff =
+                    status.getCode() != Status.Code.DEADLINE_EXCEEDED
+                    || committed == 0;
+                if (status.getCode() == Status.Code.NOT_FOUND) {
+                  future.setException(new CacheNotFoundException(digest, digestUtil));
+                } else if (!(t instanceof Exception)
+                    || !retrier.isRetriable((Exception) t)
+                    || !maybeRetry(useBackoff ? stalledBackoff.nextDelayMillis() : 0)) {
+                  future.setException(t);
                 }
               }
 
+              private boolean maybeRetry(long delayMillis) {
```

Comment:
I am wondering whether the code would look simpler if the retry logic was implemented inside a retrier?

Reply:
Yes, it did get a great deal simpler - just needed to wrap the callback executor with the context.